_posts/2024-09-09-using-search-latency-monitoring.markdown

```diff
@@ -0,0 +1,193 @@
+   - naarcha
+date: 2024-09-09 
+categories:
+  - technical-posts
+  - search
+excerpt: 
+meta_keywords:
+meta_description: 
+---
+
+# Using search latency monitoring at the coordinator node level
+
+Responsiveness is core to any search solution, and communication delays between search requests and responses can present problems for your users. This is why many administrators monitor search latency: the amount of time between when a search request is sent to an OpenSearch cluster and when a response is returned. 
+
+However, the ability to monitor search latency in OpenSearch versions earlier than 2.11 is limited. Tools such as the [Nodes Stats API](https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/) and [shard slow logs](https://opensearch.org/docs/latest/install-and-configure/configuring-opensearch/logs/#shard-slow-logs) offer latency measurements based on shard-level operations but do not provide any visibility at the coordinator node level. Because search requests often hit multiple shards simultaneously, the measured amount of time spent in a single shard search phase doesn't provide comprehensive latency information.
+
+Luckily, in OpenSearch 2.11 and later, OpenSearch offers search latency monitoring at the coordinator node level. This blog post provides information about several tools---the Node Stats API, the `phase_took` parameter, and slow logs---that can help you monitor coordinator-node-level statistics.
+
+## The coordinator node and search phases
+
+In OpenSearch, the coordinator node plays a crucial role in search execution by acting as an intermediary between the client and the cluster. When a search request is received, the coordinator node fans the request to data nodes containing the relevant shards. Depending on the type of search, the request undergoes a series of search phases, or types of search operations, initiated by the coordinator node. These phases include the following:
+
+* `dfs_pre_query` –- Depth-first search (DFS) pre-query operations
+* `query` –- Query operations
+* `fetch` –- Fetch operations
+* `dfs_query` –- DFS query operations
+* `expand` –- Expand operations
+* `can_match` –- Can match operations
+
+The following diagram shows the search phase workflow, from request to response.
+
+![Search-phases](/assets/media/blog-images/2024-09-09-using-search-latency-monitoring/search_phase2.png){:class="img-centered"}
+
+## New features for phase monitoring
+
+Before OpenSearch 2.11, OpenSearch only monitored the `query` and `fetch` shard search phases, greatly limiting the scope of search latency visibility. Now you can take advantage of the following features, some within preexisting APIs and tools, to monitor metrics for all search phases at the coordinator node level.
+
+### Nodes Stats API
+
+Introduced in OpenSearch 2.11, the [Nodes Stats API](https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/)can measure the following coordinator-node-level metrics. 
```

Comment:
[NIT] There should be a space between (https://opensearch.org/docs/latest/api-reference/nodes-apis/nodes-stats/) and can